### ISSUE TITLE
Multiple Methods for Setting JSON Markers

### ISSUE THREAD

**Issue by TetHat** [mapbox/mapbox.js#416]
I was using the method from the Custom Marker Tooltips Examples page to set my GeoJSON markerLayer -

``` js
map.markerLayer.setGeoJSON(geoJson);

map.markerLayer.on('ready', function(e) { };
```

and then using setFilter like so

``` js
map.markerLayer.setFilter(function(f) {
```

However, when trying to add filtering capability to that by using the setFilter function, map comes back undefined. But if I change my markerLayer call to the same way it is documented in the API

``` js
markerLayer = L.mapbox.markerLayer(geoJson).addTo(map);
```

and then use

``` js
markerLayer.eachLayer(function(marker) { };
```

to add in my custom content, the setFilter function works by using

``` js
markerLayer.setFilter(function(f) {
```

Wouldn't it make sense to have only one function and consolidate the way this is done? Or fix the setFilter function to work with the GeoJSON from the function -
map.markerLayer.setGeoJSON(geoJson);

Thanks.

**Comment by tmcw** [comment]
Hi - please post in [http://support.mapbox.com/](http://support.mapbox.com/) for support queries - this is a bug tracker.

**Comment by TetHat** [comment]
That would be good if I needed support. My post was saying the setGeoJSON function on this examples page
http://www.mapbox.com/mapbox.js/example/v1.0.0/custom-marker-tooltip/
is a different setGeoJSON method than the API documents﻿, and does not work with other functions such as the setFilter functions for the markers... how does that qualify as support? I fixed the problem just fine by using the ACTUAL setGeoJSON function from the API. I was just pointing out that the example page is not using the API method, and should probably be changed since other functions don't work when you use that method... I see that as something that should be fixed, not me needing support.

Thanks,

Nathan Hatfield
nathanhatfield@me.com
nathan.hatfield@bethelcollege.edu

**Comment by tmcw** [comment]
> My post was saying the setGeoJSON function on this examples page http://www.mapbox.com/mapbox.js/example/v1.0.0/custom-marker-tooltip/ is a different setGeoJSON method than the API documents

There's one function: it's [here](https://github.com/mapbox/mapbox.js/blob/v1/src/marker_layer.js#L31). It's unclear in your examples what the difference between variations would be or how you'd 'choose'.

It's also hard to parse

> map comes back undefined.

Where and when does this happen? The map variable? Or in a callback? In other code?

The rest of this issue sounds more like [the issue about ready](https://github.com/mapbox/mapbox.js/issues/405?source=cc), in which you shouldn't need to do `.on('layeradd')` when setting a literal GeoJSON object.

**Comment by TetHat** [comment]
I'm sorry, I misspoke. The function is not different, the way it is being called is different than the API. On the example page it calls the function like this:

However the API tells you to call it like this

``` js
var markerLayer = L.mapbox.markerLayer(geojson)
    .addTo(map);
```

So when I called it this way

``` js
map.markerLayer.setGeoJSON(geoJson);
```

The setFilter function did not work. I had to change it to 

``` js
var markerLayer = L.mapbox.markerLayer(geojson)
    .addTo(map);
```

in order for the setFilter function to filter the markers. If I used

map.markerLayer.setGeoJSON(geoJson);

The setFilter from the example page:

``` js
all.onclick = function() {
        food.className = '';
        this.className = 'active';
        map.markerLayer.setFilter(function(f) {
            // Returning true for all markers shows everything.
            return true;
        });
        return false;
    };
```

gave me map undefined. So now my code that works is this

``` js
var markerLayer = L.mapbox.markerLayer(geojson)
    .addTo(map);
```

and

``` js
all.onclick = function() {
        food.className = '';
        this.className = 'active';
        markerLayer.setFilter(function(f) {
            // Returning true for all markers shows everything.
            return true;
        });
        return false;
    };
```

So my thoughts were the setFilter function should be able to work with the first way it is called on the example page:

``` js
map.markerLayer.setGeoJSON(geoJson);
```

But it does not... so that is why I documented it as a bug. However, I could be wrong, and it could be intentional the setFilter function does not work when it is called this way.
Is that a better explanation? Let me know if anything else is unclear.﻿

Nathan Hatfield
nathanhatfield@me.com
nathan.hatfield@bethelcollege.edu

**Comment by tmcw** [comment]
> However the API tells you to call it like this

That is the code for _adding a new marker layer_.

> map.markerLayer.setGeoJSON(geoJson);

This is the code for changing the GeoJSON content on the 'built-in' markerLayer

> gave me map undefined. So now my code that works is this

Again, you need to be more clear. 'gave me map undefined': is this an error? a console statement? where?

**Comment by TetHat** [comment]
It came up in my Firebug console. When I had my marker layer using
map.markerLayer.setGeoJSON(geoJson);
And I clicked on a list item to filter the markers, every marker on the page disappeared because the JavaScript gave me the error map is undefined. As I have said, I have already fixed this by changing my call to

markerLayer = L.mapbox.markerLayer(geoJson).addTo(map);

The setFilter works fine on that.

So why is the
map.markerLayer.setGeoJSON(geoJson);
not something documented in the API but used in an example? Nowhere in the API for the markerLayer do I see this, explaining that it is changing the GeoJSON rather than adding it. To me that doesn't seem very useful to not have this call documented.

In the example page, it explicitly says in a comment above the call,

//Add features to the map
map.markerLayer.setGeoJSON(geoJson);

So why when I used this to initially add my GeoJSON to the map, just like in your example page, did it not work with setFilter?

And why would changing the GeoJSON compared to the adding GeoJSON not allow the setFilter function? When I combined the code from the Custom Marker Tooltip example page with the Filter Marker example page, that's when it cleared all my markers when trying to filter them. I don't see why if they both use the same setGeoJSON function, the setFilter function would work with one call but not another. And I know the code was not incorrect, because the setGeoJSON worked fine separately, and I got the code straight from your example pages. ﻿

Nathan Hatfield
nathanhatfield@me.com
nathan.hatfield@bethelcollege.edu

**Comment by tmcw** [comment]
Can you please post on the GitHub interface and use Markdown? Your replies are coming through as [one big block of text](https://github.com/mapbox/mapbox.js/issues/416#issuecomment-19395113). And please post full code for what you're trying, it's not feasible to debug these issues without context.

**Comment by TetHat** [comment]
Sorry about that. So here is what I started out with

``` javascript
map = L.mapbox.map('map', 'tetanus.map-wt0jzglg', {fullscreenControl: true})
          .setView([41.67830, -86.19366], 17);
var geoJson = { 800 lines of code that make up markers }
map.markerLayer.setGeoJSON(geoJson);
map.markerLayer.eachLayer(function(marker) {

            var feature = marker.feature;

            // Create custom popup content
            var popupContent = '<div id="gallery" class="ad-gallery">' +
                                    '<div class="ad-image-wrapper"></div><div class="ad-controls"></div>' +
                                    '<div class="ad-nav"><div class="ad-thumbs"><ul class="ad-thumb-list">' + feature.properties.images + '</ul></div></div>' +
                                    '<h2>' + feature.properties.title + '</h2>' +
                                '</div>' + 
                                '<div id="descriptions">' + feature.properties.description + '</div>' +
                                '<div id="cycle"><span onclick="if(position == 0) { position = 52; }; previous(position--);" style="z-index:999999 !important; cursor:pointer;" class="previous">Previous</span><span onclick="if(position == 51) { position = -1; }; next(position++);" style="z-index:999999 !important; cursor:pointer;" class="next">Next</span></div>';

            // http://leafletjs.com/reference.html#popup
            marker.bindPopup(popupContent,{
                closeButton: true,
                minWidth:300
            });

        });
var campus = document.getElementById('filter-campus');
campus.onclick = function(e) {
        housing.className = '';
        grocery.className = '';
        food.className = '';
        showAll.className = '';
        this.className = 'active';
        // The setFilter function takes a GeoJSON feature object
        // and returns true to show it or false to hide it.

        map.markerLayer.setFilter(function(f) {
            return f.properties['marker-color'] === '#5b8dd3';
        });
        map.markerLayer.eachLayer(function(marker) {

            var feature = marker.feature;

            // Create custom popup content
            var popupContent = '<div id="gallery" class="ad-gallery">' +
                                    '<div class="ad-image-wrapper"></div><div class="ad-controls"></div>' +
                                    '<div class="ad-nav"><div class="ad-thumbs"><ul class="ad-thumb-list">' + feature.properties.images + '</ul></div></div>' +
                                    '<h2>' + feature.properties.title + '</h2>' +
                                '</div>' + 
                                '<div id="descriptions">' + feature.properties.description + '</div>' +
                                '<div id="cycle"><span onclick="if(position == 0) { position = 52; }; previous(position--);" style="z-index:999999 !important; cursor:pointer;" class="previous">Previous</span><span onclick="if(position == 51) { position = -1; }; next(position++);" style="z-index:999999 !important; cursor:pointer;" class="next">Next</span></div>';

            // http://leafletjs.com/reference.html#popup
            marker.bindPopup(popupContent,{
                closeButton: true,
                minWidth:300
            });

        });
        return false;
    };
```

**Comment by TetHat** [comment]
So where it says map.markerLayer.setFilter inside the campus.onclick function is where it comes up with map is undefined. Here is the code I changed it to in order for it to work.

``` javascript
map = L.mapbox.map('map', 'tetanus.map-wt0jzglg', {fullscreenControl: true})
          .setView([41.67830, -86.19366], 17);
var geoJson = { 800 lines of code that make up markers }
markerLayer = L.mapbox.markerLayer(geoJson).addTo(map);
markerLayer.eachLayer(function(marker) {

            var feature = marker.feature;

            // Create custom popup content
            var popupContent = '<div id="gallery" class="ad-gallery">' +
                                    '<div class="ad-image-wrapper"></div><div class="ad-controls"></div>' +
                                    '<div class="ad-nav"><div class="ad-thumbs"><ul class="ad-thumb-list">' + feature.properties.images + '</ul></div></div>' +
                                    '<h2>' + feature.properties.title + '</h2>' +
                                '</div>' + 
                                '<div id="descriptions">' + feature.properties.description + '</div>' +
                                '<div id="cycle"><span onclick="if(position == 0) { position = 52; }; previous(position--);" style="z-index:999999 !important; cursor:pointer;" class="previous">Previous</span><span onclick="if(position == 51) { position = -1; }; next(position++);" style="z-index:999999 !important; cursor:pointer;" class="next">Next</span></div>';

            // http://leafletjs.com/reference.html#popup
            marker.bindPopup(popupContent,{
                closeButton: true,
                minWidth:300
            });

        });
var campus = document.getElementById('filter-campus');
campus.onclick = function(e) {
        housing.className = '';
        grocery.className = '';
        food.className = '';
        showAll.className = '';
        this.className = 'active';
        // The setFilter function takes a GeoJSON feature object
        // and returns true to show it or false to hide it.

        markerLayer.setFilter(function(f) {
            return f.properties['marker-color'] === '#5b8dd3';
        });
        markerLayer.eachLayer(function(marker) {

            var feature = marker.feature;

            // Create custom popup content
            var popupContent = '<div id="gallery" class="ad-gallery">' +
                                    '<div class="ad-image-wrapper"></div><div class="ad-controls"></div>' +
                                    '<div class="ad-nav"><div class="ad-thumbs"><ul class="ad-thumb-list">' + feature.properties.images + '</ul></div></div>' +
                                    '<h2>' + feature.properties.title + '</h2>' +
                                '</div>' + 
                                '<div id="descriptions">' + feature.properties.description + '</div>' +
                                '<div id="cycle"><span onclick="if(position == 0) { position = 52; }; previous(position--);" style="z-index:999999 !important; cursor:pointer;" class="previous">Previous</span><span onclick="if(position == 51) { position = -1; }; next(position++);" style="z-index:999999 !important; cursor:pointer;" class="next">Next</span></div>';

            // http://leafletjs.com/reference.html#popup
            marker.bindPopup(popupContent,{
                closeButton: true,
                minWidth:300
            });

        });
        return false;
    };
```

By doing this, I removed the map.markerLayer in the setFilter call, and made it just markerLayer. Now it works and filters my markers as I expected the first section of code to do.
